### PR TITLE
Migrate ownable

### PR DESCRIPTION
**Pull request by andrew-fleming**
Resolves (partial) #548.

Follows #584 with `#[internal]` macros for consistency.

Note: error messages do not include the `Ownable:` prefix because they will exceed the 31 character limit.

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/openzeppelin/access/ownable.cairo [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        _owner: ContractAddress
+    }
+
+    #[event]
+    fn OwnershipTransferred(previous_owner: ContractAddress, new_owner: ContractAddress) {}
+
+    #[internal]
+    fn initializer() {
+        let caller: ContractAddress = get_caller_address();
+        _owner::write(caller);
```
Comment (martriay) [inline src/openzeppelin/access/ownable.cairo:19]:
```suggestion
        _transfer_ownership(caller);
```
we want to emit the event, no?

File: src/openzeppelin/access/ownable.cairo [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    #[internal]
+    fn assert_only_owner() {
+        let owner: ContractAddress = _owner::read();
+        let caller: ContractAddress = get_caller_address();
+        assert(!caller.is_zero(), 'Caller is the zero address');
```
Comment (martriay) [inline src/openzeppelin/access/ownable.cairo:26]: why were we prohibiting this?

the only scenario I can think of where `caller == 0` is in `__execute__`. i guess it would be weird to use this guard in that context considering that the caller is always known (it's `0`)

File: src/openzeppelin/access/ownable.cairo [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    #[internal]
+    fn owner() -> ContractAddress {
+        _owner::read()
+    }
+
+    #[internal]
+    fn transfer_ownership(new_owner: ContractAddress) {
+        assert(!new_owner.is_zero(), 'New owner is zero address');
```
Comment (martriay) [inline src/openzeppelin/access/ownable.cairo:37]:
```suggestion
        assert(!new_owner.is_zero(), 'New owner is the zero address');
```
closer to [OZ Contracts for solidity's error message](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol#L70) (still lacking the `Ownable:` we can't add due max lenght)

File: src/openzeppelin/access/ownable.cairo [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#[contract]
+mod Ownable {
+    use starknet::ContractAddress;
+    use starknet::ContractAddressZeroable;
+    use starknet::contract_address_const;
```
Comment (martriay) [inline src/openzeppelin/access/ownable.cairo:5]:
```suggestion
```
unused maybe?

File: src/openzeppelin/tests/test_ownable.cairo [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+fn test_renounce_ownership() {
+    setup();
+    Ownable::renounce_ownership();
+    assert(Ownable::owner() == ZERO(), 'Should renounce ownership');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('Caller is the zero address', ))]
+fn test_renounce_ownership_from_zero_address() {
```
Comment (martriay) [inline src/openzeppelin/tests/test_ownable.cairo:69]: should we also test transferring ownership from zero, since we're prohibiting it?

**Review by martriay — COMMENTED** [review]
overall pretty good, left some comments!

File: src/openzeppelin/access/ownable.cairo (hunk 2 above)
Comment (andrew-fleming) [reply src/openzeppelin/access/ownable.cairo:26]: I believe the concern was for scenarios where if an owner renounces (setting owner to `0`), another actor couldn't call protected functions as `0`

Comment (martriay) [reply src/openzeppelin/access/ownable.cairo:25]: you meant "could" as in not having the check would open that door, right? but that would only be possible if `assert_only_owner` was being called in `__execute__` for which they should have been able to bypass `__validate__` first.

the other possibility is that `__validate__` uses `assert_only_owner` but again, the developer should be aware by then that the caller will ALWAYS be zero in that context. i'm not sure this is an actual risk.

Comment (andrew-fleming) [reply src/openzeppelin/access/ownable.cairo:25]: Good points. I'll remove the check

Comment (martriay) [reply src/openzeppelin/access/ownable.cairo:25]: just out of extra caution would like to summon @Amxx and @frangio, considering our "make it extremely hard to misuse" design principle

Comment (frangio) [reply src/openzeppelin/access/ownable.cairo:25]: This doesn't seem to be about "hard to misuse", but about "defense in depth". Without much context I would keep this check.

File: src/openzeppelin/access/ownable.cairo [hunk 6]
```diff
@@ -1,9 +1,6 @@
 #[contract]
 mod Ownable {
-    use starknet::ContractAddress;
-    use starknet::ContractAddressZeroable;
-    use starknet::contract_address_const;
-    use starknet::get_caller_address;
+    use starknet::{ContractAddress, contract_address::ContractAddressZeroable, get_caller_address};
```
Comment (martriay) [inline src/openzeppelin/access/ownable.cairo:3]: if this is not the linter, i vote for separate lines

Comment (martriay) [reply src/openzeppelin/access/ownable.cairo:3]: also couldn't simply be `ContractAddressZeroable`, without `contract_address::`?

**Review by martriay — APPROVED** [review]
looks good! left a style comment, otherwise good to merge

File: src/openzeppelin/access/ownable.cairo (hunk 6 above)
Comment (martriay) [reply src/openzeppelin/access/ownable.cairo:3]: what do you think @ericnordelo?

Comment (ericnordelo) [reply src/openzeppelin/access/ownable.cairo:3]: `ContractAddressZeroable` has been removed from the scope of `starknet`, so we can't use it like that, but the reason they removed it is that we don't need to import it anymore. It is enough just importing the trait instead (`use zeroable::Zeroable;`)

```suggestion
    use starknet::ContractAddress;
    use starknet::get_caller_address;
```

File: src/openzeppelin/tests/test_ownable.cairo [hunk 7]
```diff
@@ -0,0 +1,88 @@
+use openzeppelin::access::ownable::Ownable;
+
+use starknet::{ContractAddress, contract_address_const, testing::set_caller_address};
```
Comment (ericnordelo) [inline src/openzeppelin/tests/test_ownable.cairo:3]: Even when in this case is just a function, we would be using `testing` module members many times, then following Rust best practices, I think it will be better for clarity to keep the convention of importing the module instead of the functions directly (the multiline imports could be another convention, with the exception of particularly related members, like `add` and a `mul` in an `operations` module)

```suggestion
use starknet::ContractAddress;
use starknet::contract_address_const;
use starknet::testing;
```

File: src/openzeppelin/token/erc20.cairo [hunk 8]
```diff
@@ -15,11 +15,9 @@ trait IERC20 {
 #[contract]
 mod ERC20 {
     use super::IERC20;
-    use starknet::get_caller_address;
-    use starknet::ContractAddress;
-    use starknet::ContractAddressZeroable;
-    use zeroable::Zeroable;
     use integer::BoundedInt;
+    use starknet::{ContractAddress, contract_address::ContractAddressZeroable, get_caller_address};
```
Comment (ericnordelo) [inline src/openzeppelin/token/erc20.cairo:19]:
```suggestion
    use starknet::ContractAddress;
    use starknet::get_caller_address;
```

**Review by ericnordelo — COMMENTED** [review]
Looking good! Left just a couple of code format suggestions.

File: src/openzeppelin/tests/test_ownable.cairo (hunk 7 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_ownable.cairo:3]: Sounds good to me

**Comment by dic0de** [comment]
Hello, Quick question, why did you decide to use the `#[internal]` macros to include in the `ownable` contracts. I wanted to create an issue thinking it was an error, but it seems deliberate. I would wish to know your reasons if you do not mind.

**Comment by martriay** [comment]
@dic0de why do you think it's an error?

**Comment by dic0de** [comment]
Because here https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-1/src/openzeppelin/access/ownable.cairo, every function from `initializer` is marked `#[internal]` including functions that only perform read actions, hence my question. I just would wish to know why you implemented it that way out of curiosity.

**Comment by martriay** [comment]
the `#[internal]` annotation does nothing, it's a convention we started to indicate a function is meant to be internal and not external, just that. Despite not being an actual compiler annotation I still don't see what would be the problem or apparent mistake.
